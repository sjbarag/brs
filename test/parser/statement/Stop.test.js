const brs = require("brs");
const { Lexeme } = brs.lexer;
const { Int32 } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("stop statement", () => {
    it("cannot be used as a local variable", () => {
        let { statements, errors } = brs.parser.Parser.parse([
            token(Lexeme.Stop, "stop"),
            token(Lexeme.Equal, "="),
            token(Lexeme.True, "true"),
            EOF,
        ]);

        //should be an error
        expect(errors).toHaveLength(1);
        expect(statements).toBeDefined();
        expect(statements).not.toBeNull();
        expect(statements).toMatchSnapshot();
    });

    it("is valid as a statement", () => {
        let { statements, errors } = brs.parser.Parser.parse([token(Lexeme.Stop, "stop"), EOF]);
        expect(errors[0]).toBeUndefined();
        expect(statements).toMatchSnapshot();
    });

    it("can be used as an object property", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                theObject = {
                    stop: false
                }
                theObject.stop = true
            end sub
        `);
        let { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });
});
