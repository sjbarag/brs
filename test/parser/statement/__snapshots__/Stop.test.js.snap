// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stop statement can be used as an object property 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 7,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "statements": Array [
          Assignment {
            "name": Object {
              "isReserved": false,
              "kind": "Identifier",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 25,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 16,
                  "line": 3,
                },
              },
              "text": "theObject",
            },
            "tokens": Object {
              "equals": Object {
                "isReserved": false,
                "kind": "Equal",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 27,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 26,
                    "line": 3,
                  },
                },
                "text": "=",
              },
            },
            "type": "Assignment",
            "value": AALiteral {
              "close": Object {
                "isReserved": false,
                "kind": "RightBrace",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 17,
                    "line": 5,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 5,
                  },
                },
                "text": "}",
              },
              "elements": Array [
                Object {
                  "name": BrsString {
                    "kind": 3,
                    "value": "stop",
                  },
                  "value": Literal {
                    "_location": Object {
                      "end": Object {
                        "column": 31,
                        "line": 4,
                      },
                      "file": "",
                      "start": Object {
                        "column": 26,
                        "line": 4,
                      },
                    },
                    "type": "Literal",
                    "value": BrsBoolean {
                      "kind": 2,
                      "value": false,
                    },
                  },
                },
              ],
              "open": Object {
                "isReserved": false,
                "kind": "LeftBrace",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 29,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 28,
                    "line": 3,
                  },
                },
                "text": "{",
              },
              "type": "AALiteral",
            },
          },
          DottedSet {
            "name": Object {
              "isReserved": true,
              "kind": "Identifier",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 30,
                  "line": 6,
                },
                "file": "",
                "start": Object {
                  "column": 26,
                  "line": 6,
                },
              },
              "text": "stop",
            },
            "obj": Variable {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 25,
                    "line": 6,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 6,
                  },
                },
                "text": "theObject",
              },
              "type": "Variable",
            },
            "type": "DottedSet",
            "value": Literal {
              "_location": Object {
                "end": Object {
                  "column": 37,
                  "line": 6,
                },
                "file": "",
                "start": Object {
                  "column": 33,
                  "line": 6,
                },
              },
              "type": "Literal",
              "value": BrsBoolean {
                "kind": 2,
                "value": true,
              },
            },
          },
        ],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 7,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 7,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
    "type": "Stmt_Function",
  },
]
`;

exports[`stop statement cannot be used as a local variable 1`] = `
Array [
  Stop {
    "tokens": Object {
      "stop": Object {
        "isReserved": true,
        "kind": "Stop",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "stop",
      },
    },
    "type": "Stop",
  },
]
`;

exports[`stop statement is valid as a statement 1`] = `
Array [
  Stop {
    "tokens": Object {
      "stop": Object {
        "isReserved": true,
        "kind": "Stop",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "stop",
      },
    },
    "type": "Stop",
  },
]
`;
